Prevent downtimes scheduled within 24 hours from being deleted

ElliottKasoar opened the pull request:
Resolves #269

Added a check when attempting to delete a scheduled downtime to prevent deletion if the start time is within 24 hours.

As this check is also carried out for the old and new start times when editing a downtime, this functionality has been factored out into `validateStartTime`.

The additional validation when deleting a downtime has been factored out with the existing check (if the downtime is on-going) for clarity.

ElliottKasoar commented:
(No new changes, I squashed the very minor commits)

ineilson reviewed (approved):
Personally, I don't like programs reporting using contractions - "can't" - but use here pre-dates these changes.
Similarly for using short (<2 character) variable names, which angers the 'standard' PHPMD ruleset.
Does this address the issue of logging a failed delete attempt raised in #269?
Otherwise, looks good.

ElliottKasoar commented:
> Personally, I don't like programs reporting using contractions - "can't" - but use here pre-dates these changes.
> Similarly for using short (<2 character) variable names, which angers the 'standard' PHPMD ruleset.

I can definitely update the exception message, as I've changed it a bit anyway. I did consider changing `$dt`, but it appears 50+ times within Downtime.php, as well as in other scripts, so it would require a lot of changes to maintain consistent naming, which I imagine is a bigger change than we want here?

> Does this address the issue of logging a failed delete attempt raised in #269?

This was discussed briefly with @gregcorbett when I started work on the issue. I think we felt that it made sense to treat the error in the same way as [downtimes that have already started](https://github.com/GOCDB/gocdb/blob/dev/lib/Gocdb_Services/Downtime.php#L665), which wouldn't allow an HTTP 400 to be issued.

(Following discussions with @ineilson) There is (currently) no WriteAPI function to manipulate downtimes, so the above isn't an issue. An alternative would be to instead flag an attempted (authorised) deletion with some indication on the web display, but that would be a more substantial change.

ineilson commented:
#241 and the referenced GGUS ticket may be relevant here if any future work is planned.

ineilson commented:
> I can definitely update the exception message, as I've changed it a bit anyway. I did consider changing $dt, but it appears 50+ times within Downtime.php, as well as in other scripts, so it would require a lot of changes to maintain consistent naming, which I imagine is a bigger change than we want here?

The Big Boss @gregcorbett can decide whether 2 char len variable $dt is included in #265, along with $em, $dn and $id.

gregcorbett commented:
> The Big Boss @gregcorbett can decide whether 2 char len variable $dt is included in #265, along with $em, $dn and $id.

I would say changing `$dt` to something more descriptive throughout the code should be it own separate change in the future.

ElliottKasoar commented:
> > Personally, I don't like programs reporting using contractions - "can't" - but use here pre-dates these changes.
> 
> I can definitely update the exception message, as I've changed it a bit anyway.

I removed contractions from exceptions/comments where I've made changes, and squashed into existing commits.

gregcorbett commented:
Just rebasing this on top of the latest dev
